sw/device/tests/crypto/kdf_kmac_functest.c

```diff
@@ -0,0 +1,112 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/drivers/entropy.h"
+#include "sw/device/lib/crypto/drivers/kmac.h"
+#include "sw/device/lib/crypto/impl/integrity.h"
+#include "sw/device/lib/crypto/include/datatypes.h"
+#include "sw/device/lib/crypto/include/hash.h"
+#include "sw/device/lib/crypto/include/kdf.h"
+#include "sw/device/lib/crypto/include/mac.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+// The autogen rule that creates this header creates it in a directory named
+// after the rule, then manipulates the include path in the
+// cc_compilation_context to include that directory, so the compiler will find
+// the version of this file matching the Bazel rule under test.
+#include "kdf_testvectors.h"
+
+// Global pointer to the current test vector.
+static kdf_kmac_test_vector_t *current_test_vector = NULL;
+
+/**
+ * Get the mode for KMAC based on test operation.
+ *
+ * @param test_operation Security strength of KMAC.
+ * @param[out] mode KMAC mode enum value.
+ */
+status_t get_kmac_mode(kdf_test_operation_t test_operation,
+                       otcrypto_kmac_mode_t *mode) {
+  switch (test_operation) {
+    case kKdfTestOperationKmac128:
+      *mode = kOtcryptoKmacModeKmac128;
+      break;
+    case kKdfTestOperationKmac256:
+      *mode = kOtcryptoKmacModeKmac256;
+      break;
+    default:
+      LOG_INFO("Invalid test operation is given for KDF-KMAC");
+      return INVALID_ARGUMENT();
+  }
+  return OK_STATUS();
+}
+
+/**
+ * Run the test pointed to by `current_test_vector`.
+ */
+static status_t run_test_vector(void) {
+  // Below, `km` prefix refers to keying_material, and
+  // `kdk` prefix refers to key derivation key
+  size_t km_key_length = current_test_vector->keying_material.config.key_length;
+
+  size_t km_num_words = km_key_length / sizeof(uint32_t);
+  if (km_key_length % sizeof(uint32_t) != 0) {
+    km_num_words++;
+  }
+  uint32_t km_buffer[km_num_words];
+
+  otcrypto_kmac_mode_t mode;
+  TRY(get_kmac_mode(current_test_vector->test_operation, &mode));
+
+  otcrypto_blinded_key_t keying_material = {
+      .config =
+          {
+              .key_length = km_key_length,
+              .hw_backed = kHardenedBoolFalse,
+              .security_level = kOtcryptoKeySecurityLevelLow,
+          },
```

Comment:
This is going to be an invalid configuration, since e.g. `key_mode` and `.exportable` are not set.

Reply:
Does that mean KDF-KMAC driver needs to check `.exportable` and `.key_mode` fields? Should I always validate `key_mode` by passing it through `keyblob_ensure_xor_masked`?

Reply:
Since `keying_material` is the generated key, it could be anything from `otcrypto_key_mode_t`. Do you think it makes sense to have some util functions for validating config field? Especially, in the case of KDF, it might be good to check if `key_mode` and `key_length` fields match each other.

Reply:
I don't think the KDF-KMAC code actually needs to check those extra values, but I think it's good practice in tests to use fully populated key structs just in case the code changes and does check those values, just for maintainability -- it might lead to weird error messages down the line otherwise.